packages/clay-labels-input-field/src/ClayLabelsInputField.js

```diff
@@ -0,0 +1,427 @@
+
+		this._removeFocusedLabel();
+
+		switch (event.data) {
+		case ',':
+			return this._handleLabelAdded(event);
+		default:
+			return !this.emit({
+				data: {
+					value,
+				},
+				name: 'queryChange',
+				originalEvent: event,
+			});
+		}
+	}
+
+	/**
+	 * Handles form interactions and propagates corresponding events.
+	 * @param {!Event} event
+	 * @protected
+	 * @return {Boolean} If the event has been prevented or not.
+	 */
+	_handleOnKeydown(event) {
+		switch (event.key) {
+		case 'Enter':
+			event.preventDefault();
+			if (this._labelFocused) {
+				return this._handleLabelRemoved(this._labelFocused);
+			} else {
+				return this._handleLabelAdded(event);
+			}
+		case 'Backspace':
+			if (!this.refs.input.value) {
+				if (!this._labelFocused) {
+					return this._handleLabelFocus(event);
+				} else {
+					return this._handleLabelRemoved(this._labelFocused);
+				}
+			}
+			break;
+		case 'ArrowLeft':
+			if (!this.refs.input.value && this._labelFocused) {
+				return this._handleLabelFocus(event, true);
+			}
+			break;
+		case 'ArrowRight':
+			if (!this.refs.input.value && this._labelFocused) {
+				return this._handleLabelFocus(event, false);
+			}
+			break;
+		case 'ArrowUp':
+			this._setFocusItemDropdown(true);
+			break;
+		case 'ArrowDown':
+			this._setFocusItemDropdown(false);
+			break;
+		}
+	}
+
+	/**
+	 * Removes the focus from the focused element.
+	 * @protected
+	 */
+	_removeFocusedLabel() {
+		if (this._labelFocused) {
+			this._labelFocused.classList.remove('label-focus');
+			this._labelFocused = null;
+		}
+	}
+
+	/**
+	 * Handle the interactions in the dropdown and add focus on the items.
+	 * @protected
+	 * @param {!Boolean} direction
+	 */
+	_setFocusItemDropdown(direction) {
+		if (this.filteredItems.length) {
+			const elements = this.refs.dropdown.querySelectorAll(
+				'a[id="item"]'
+			);
+
+			if (direction && this._dropdownItemFocused === 0) {
+				this.refs.input.focus();
+				this._dropdownItemFocused = null;
+			} else {
+				this._dropdownItemFocused =
+					this._dropdownItemFocused === null ||
+					elements.length - 1 === this._dropdownItemFocused
+						? 0
+						: direction
+							? this._dropdownItemFocused - 1
+							: this._dropdownItemFocused + 1;
+
+				elements[this._dropdownItemFocused].focus();
+			}
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	attached() {
+		this._labelFocused = null;
+		this._dropdownItemFocused = null;
+	}
+
+	/**
+	 * Clears the input value.
+	 * @public
+	 */
+	clearInput() {
+		this.refs.input.value = '';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	created() {
+		this._eventHandler = new EventHandler();
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	disposed() {
+		this._eventHandler.removeAllListeners();
+	}
+
+	/**
+	 * Helper method to filter a list based on a string.
+	 * @param {!Array<string>} data
+	 * @param {!string} query
+	 * @public
+	 * @return {Array} A list of items containing the corresponding characters
+	 */
+	filterAutocomplete(data, query) {
```

Comment:
Let's extract this to something like `clay-dataprovider`. See [DataProvider](https://github.com/liferay/clay/pull/971/files#diff-fc5f1ca3d00e2b11a1a6029e3b277473R24) for some reference or [Kendo DataSource](https://demos.telerik.com/kendo-ui/datasource/index).

We should support both local and remote data. See also how we started doing it in [clay-charts-shared](https://github.com/liferay/clay/blob/master/packages/clay-charts-shared/src/data.js) and [ClayBase](https://github.com/liferay/clay/blob/master/packages/clay-charts/src/ChartBase.js)

Reply:
I was really thinking about not dealing with the DataProvider here 😅, I'll separate a task to just deal with it. Thanks for the examples.

Reply:
Okay, let's get this component through and open an issue to abstract that!